### ISSUE TITLE
Master branch can't open large Tif files

### ISSUE THREAD

**Issue by Andrea-virology** [ilastik/ilastik#1193]
Hi everyone, 
I'm using the current master branch. I can open small tif files (3 channels, 100 time frames) nicely. However, I could not open a large Tif file of 8.4 GB. I then get the message invalid TIFF image index /miniconda/conda-bld/work-src/impex/tiff.cxx:356
Here is the log file.

Perhaps something can be done about it and be directly included in the next release.

Thanks
[ilastik_log.txt](https://github.com/ilastik/ilastik/files/167501/ilastik_log.txt)

**Comment by stuarteberg** [comment]
Thanks for the bug report.  Can you confirm that there are no other differences (beyond just the number of pages) between the two files you tested with?

The specific error you're seeing comes from `vigra/src/impex/tiff.cxx`:
```C++
    void TIFFDecoderImpl::init(unsigned int imageIndex)
    {
        // set image directory, if necessary:
        if (imageIndex != TIFFCurrentDirectory(tiff))
        {
            if (!TIFFSetDirectory( tiff, (tdir_t)(imageIndex)) )
                vigra_fail( "Invalid TIFF image index" );
        }

```

I'm not familiar enough with the libtiff API to know why `TIFFSetDirectory` would fail for an image with a lot of pages.  @ukoethe, do you have any theories?

**Comment by Andrea-virology** [comment]
There is no other difference. I generated the small file by duplicating the first 100 frames of the large one. When I duplicated the whole 8.4 GB file and saved it exactly the same way as the small one, I couldn't open it in Ilastik. So the files were generated the same way and saved the same way. I conclude that the size is the only difference.

**Comment by ukoethe** [comment]
Can you create two other versions of your test file where the number of frames is chosen such that the files are just under 2 GB and 4GB respectively? If it works for 2 GB (and maybe 4GB) but not 8 GB, the TIFF library is perhaps compiled without support for files exceeding 2<sup>32</sup> bytes.

**Comment by ukoethe** [comment]
@stuarteberg @ilastikdev  I supsect that structured learning hasn't been synchronized with the latest changes in master, because the log reports the error in 
```python
page_data = vigra.impex.readImage(self._filepath, dtype='NATIVE', 
                                  index=tiff_page_list_index, order='C')
```
which doesn't contain my recent fix `index=int(tiff_page_list_index)`.

**Comment by ilastikdev** [comment]
To clarify, this error happened in the pixel classification workflow using all current master branches of lazyflow, volumina, and ilastik when we installed the development system.

The date escapes me now, but it could have been before
the commit 86ee498c8ea0c682c21031c38ce265d457aa8430 made it into master
(commit was made 14 days ago).

I will only be able to look into this next week when Virology personnel is back in the building.

**Comment by Andrea-virology** [comment]
Hi everyone, 
Janez and I checked that the master branch we used has the current tif fix, but it still did not work. I tried different file sizes. Loading data in pixel classification works below 2 GB and above 2GB. However, above 4 GB import fails. So I created a dropbox link with such a file over 4GB that does not load. Perhaps it helps you to figure out the problem. Please let me know when I can delete it, because it takes most of the space of my dropbox account.
https://www.dropbox.com/sh/890dk6qy4hw1l6d/AAC5u4upnj0Dr0RGctlDjVYQa?dl=0

**Comment by ukoethe** [comment]
Could it simply be an out-of-memory error that's not properly reported?

**Comment by ukoethe** [comment]
OK, I got the file and can reproduce the error on Linux (didn't try Windows, because I'll probably run out of memory).

**Comment by ukoethe** [comment]
@Andrea-virology ilastik thinks the dataset is (800, 3, 1002, 1004) as "tcyx". Is this correct?

**Comment by Andrea-virology** [comment]
yes, this is correct.

**Comment by ukoethe** [comment]
The problem is that the file is not actually TIFF, but ImageJ's private flavor of the format. It seems that everyone is messing around with the TIFF format at will, adding new attributes and re-interpreting the file contents like crazy. The `tifffile` library (which reads the metadata that are displayed in the data selection applet) supports about eight flavors of TIFF, and as it happens this includes ImageJ TIFF. However, the code that actually imports the data (which eventually calls libtiff, the official TIFF implementation) only supports real TIFF and sees only the first frame. Therefore, it fails when more than one frame is requested.

The easiest solution is to convert the data to a proper format, preferably HDF5.

**Comment by stuarteberg** [comment]
>The problem is that the file is not actually TIFF, but ImageJ's private flavor of the format.

Sadly, for a typical biologist, I suspect ImageJ TIFF is the "real" format, regardless of what `libtiff` supports.  Is it clear why `libtiff` isn't working in this case?

Moreover, I suppose we could fall back to using `tifffile` to read the data in cases when `libtiff` doesn't work, as long as the file doesn't use JPEG compression.  Is there an easy way to determine whether or not `libtiff` can handle a given file?

**Comment by ukoethe** [comment]
> Sadly, for a typical biologist, I suspect ImageJ TIFF is the "real" format

Yeah, but they shouldn't call it TIFF then, and publish a proper spec (google hasn't come up with anything useful so far, probably we must reverse engineer the Java sources). `libtiff` doesn't work because the official header (which still exists) lies about the image contents. ImageJ writes its own header somewhere else, and God knows where it puts the data.

> I suppose we could fall back to using `tifffile` to read the data

As far as I can tell, `tifffile` can only read the header information, not the contents. `tifffile.imread()` and `vigra.readImage()` import exactly the same data (namely the first channel of the first frame).

**Comment by stuarteberg** [comment]
Related: We now support the "BigTiff" format...

**Comment by vzickus** [comment]
@stuarteberg I still run into 
```

 Invalid TIFF image index
 (/opt/conda/conda-bld/vigra_1548669310766/work/src/impex/tiff.cxx:356)
```

error while loading even a single .ome.tiff file

**Comment by JulianIST** [comment]
I also still encounter the same issue - I have a txy dataset which was modified via FIJI (so a "FIJI tiff") and Ilastik can not switch to any other tiemframe nor can it export predictions. 

Reducing the size to 3.9 GB (cutting off frames) removes the issue. 

>  Invalid TIFF image index
 (Users\IEUser\miniconda3\conda-bld\vigra_1600359360699\work\src\impex\tiff.cxx:356)

**Comment by oanegros** [comment]
Bumping this, as it seems to still be an issue for me.
Fiji's implementation of tifs are reasonably industry-standard, and h5 is better, but not as universally adopted. 
Have there been any developments on this front? And could there just be a fallback option to one of the supported tifreaders (such as `tifffile`) see also https://forum.image.sc/t/reading-imagej-large-tif-file-format-in-other-software/63757/6

**Comment by k-dominik** [comment]
still an issue, after [#2729 ](https://github.com/ilastik/ilastik/pull/2733)
